csrc/val_graph.h

```diff
@@ -0,0 +1,216 @@
+//
+// - {I0, I2, I4}
+// - {I1, I3, I5}
+// - {I2*I3, I4*I5}
+// - {I2*I3/4, I4*I5/4}
+// - {4, 4}
+//
+// and ExprGroups of:
+//
+// - {merge of I2 and I3, merge of I4 and I5}
+// - {split of I2*I3, split of I4*I5}
+//
+// ValGraph can be used with any Val types, however, it's currenty
+// only tested with IterDomain. Some of the routines might need to be
+// extended for other Val types.
+
+using ValGroup = std::shared_ptr<VectorOfUniqueEntries<Val*>>;
+using ValGroups = VectorOfUniqueEntries<ValGroup>;
+using ExprGroup = std::shared_ptr<VectorOfUniqueEntries<Expr*>>;
+using ExprGroups = VectorOfUniqueEntries<ExprGroup>;
+
+class ValGraph {
+ public:
+  ValGraph() = default;
+
+  ValGraph(const ValGraph& other);
+  ValGraph(ValGraph&& other) = default;
+
+  ValGraph& operator=(const ValGraph& other);
+  ValGraph& operator=(ValGraph&& other) = default;
+
+  ValGraph(bool propagate_through_exprs)
+      : propagate_through_exprs_(propagate_through_exprs) {}
+
+  // Returns the disjoint val set.
+  const DisjointSets<Val*>& disjointValSets() const {
+    return disjoint_vals_;
+  }
+
+  // Returns the disjoint Expr set.
+  const DisjointSets<Expr*>& disjointExprSets() const {
+    return disjoint_exprs_;
+  }
+
+  // Return if there's a group entry in the graph for this expr
+  bool hasGroup(Expr* expr) const;
+
+  // Return if there's a group entry in the graph for this val
+  bool hasGroup(Val* val) const;
+
+  // Convert expr to its exprGroup, assert that it exists.
+  const ExprGroup& toGroup(Expr* expr) const;
+
+  // Convert Val to its ValGroup, assert that it exists.
+  const ValGroup& toGroup(Val* val) const;
+
+  // Return output/input Val groups of provided expr
+  // Note that the same ValGroup can show up multiple times, so the
+  // output type cannot be VectorOfUniqueEntries
+  std::vector<ValGroup> outputGroups(const ExprGroup& expr) const;
+  std::vector<ValGroup> inputGroups(const ExprGroup& expr) const;
+
+  // Recursively traverses uses of the IdGroups in 'of' and returns all
+  // ExprGroups that have a use in their definition of provided of IdGroups.
+  ExprGroups allUsesOf(const ValGroups& of) const;
+
+  // Recursively traverses definitions of the IdGroups in 'of' and returns all
+  // ExprGroups used in this history of defining the 'of' IdGroups.
+  ExprGroups allDefinitionsOf(const ValGroups& of) const;
+
+  //! Returns the pointer to expressions associated with the
+  //! definitions of the provided ValGroup in the provided
+  //! mapping mode (if it exists). Nullptr is returned otherwise.
+  //!
+  //! The returned pointer is to a vector of vector of expressions. The
+  //! inner vector is proven to be equivalent based on the provided mode. The
+  //! outer vector are expression groups that are not equivalent based on the
+  //! provided mode, but produce one of the ValGroups within the same disjoint
+  //! Val set based on the provided mode.
+  const ExprGroups* getDefinitions(const ValGroup& id_group) const;
```

Comment:
nit:
```suggestion
  const ExprGroups* getDefinitions(const ValGroup& of) const;
```

Reply:
I'd avoid just using `of` as it tends to get confusing whether it's ValGroup, ValGroups, ExprGroup, or ExprGroups.

Renamed to `val_group`